amaranth/hdl/_ast.py

```diff
@@ -399,222 +440,595 @@ def __init__(self, *, src_loc_at=0):
+
+        Raises
+        ------
+        :exc:`TypeError`
+            If :pc:`other` is signed.
+        """
+        other = Value.cast(other)
+        other.__check_shamt()
+        return Operator("<<", [self, other], src_loc_at=1)
+
+    def __rlshift__(self, other):
+        """Left shift by variable amount, :pc:`other << self`.
+
+        See :mod:`__lshift__`.
+        """
+        self.__check_shamt()
+        return Operator("<<", [other, self])
+
+    def shift_left(self, amount):
+        """Left shift by constant amount.
+
+        If :pc:`amount < 0`, performs the same operation as :pc:`self.shift_right(-amount)`.
+
+        Returns
+        -------
+        :class:`Value`, :pc:`unsigned(len(self) + amount)`
+            If :pc:`self` is unsigned.
+        :class:`Value`, :pc:`signed(len(self) + amount)`
+            If :pc:`self` is signed.
+        """
+        if not isinstance(amount, int):
+            raise TypeError(f"Shift amount must be an integer, not {amount!r}")
+        if amount < 0:
+            return self.shift_right(-amount)
+        if self.shape().signed:
+            return Cat(Const(0, amount), self).as_signed()
+        else:
+            return Cat(Const(0, amount), self) # unsigned
+
+    def rotate_left(self, amount):
+        """Left rotate by constant amount.
+
+        If :pc:`amount < 0`, performs the same operation as :pc:`self.rotate_right(-amount)`.
+
+        Returns
+        -------
+        :class:`Value`, :pc:`unsigned(len(self))`, :ref:`assignable <lang-assignable>`
+        """
+        if not isinstance(amount, int):
+            raise TypeError(f"Rotate amount must be an integer, not {amount!r}")
+        if len(self) != 0:
+            amount %= len(self)
+        return Cat(self[-amount:], self[:-amount]) # meow :3
+
+    @_overridable_by_reflected("__rrshift__")
+    def __rshift__(self, other):
+        """Right shift by variable amount, :pc:`self >> other`.
+
+        Returns
+        -------
+        :class:`Value`, :pc:`unsigned(len(self))`
+            If :pc:`self` is unsigned.
+        :class:`Value`, :pc:`signed(len(self))`
+            If :pc:`self` is signed.
+
+        Raises
+        ------
+        :exc:`TypeError`
+            If :pc:`other` is signed.
+        """
+        other = Value.cast(other)
+        other.__check_shamt()
+        return Operator(">>", [self, other], src_loc_at=1)
+
+    def __rrshift__(self, other):
+        """Right shift by variable amount, :pc:`other >> self`.
+
+        See :mod:`__rshift__`.
+        """
+        self.__check_shamt()
+        return Operator(">>", [other, self])
+
+    def shift_right(self, amount):
+        """Right shift by constant amount.
+
+        If :pc:`amount < 0`, performs the same operation as :pc:`self.left_right(-amount)`.
+
+        Returns
+        -------
+        :class:`Value`, :pc:`unsigned(len(self) - amount)`
+            If :pc:`self` is unsigned.
+        :class:`Value`, :pc:`signed(len(self) - amount)`
```

Comment:
... uhhhh

```
>>> Signal(signed(10).shift_right(11)).shape()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Shape' object has no attribute 'shift_right'
```

Reply:
Um, `signed(10)` is not a value?

Reply:
I get

```
>>> Signal(signed(10)).shift_right(11).shape()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/whitequark/Projects/amaranth/amaranth/hdl/_ast.py", line 1509, in shape
    return Shape(a_shape.width, True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/whitequark/Projects/amaranth/amaranth/hdl/_ast.py", line 63, in __init__
    raise TypeError(f"Width of a signed value must be a positive integer, not {width}")
TypeError: Width of a signed value must be a positive integer, not 0
```

which does seem vaguely like a bug, or underspecification at least.